Audio field

Jbonnett opened the pull request:
see the audio tag in https://www.pivotaltracker.com/projects/253951 for more details.

Commit:
first stab at a audio field

Commit:
first stab at tests for audio Field

Commit:
added dep on the mutagen-flo branch

Commit:
changes

Commit:
audio moves closer to some semblance of function

Commit:
Working test of extracting the metadata from a mp3 file

Commit:
white space/pep8

Commit:
added a ogg test/test file

Commit:
typo got commited

Commit:
widget renders, tries to pull the more complicated widget if
armstrong.apps.audio is installed. Not really happy with how its
doing it, so i will probably circle back once i have the widget working
in the app.

Commit:
todo: not suck so badly

Commit:
Making mutagen optional

Commit:
broken commit of wip
trying to get armstrong.utils.backends to install properly so that i can
test this.

Commit:
bug fixes

Commit:
audio field now uses a generic backend

the options are mutagen-flo or id3reader
id3reader is included as the license seems very permisive,

Commit:
added the stand in audio model to the test admin site

Commit:
model fields now pre populate if they exist

its currently overzellous and will prepop on any metadata field

Commit:
prepopulation and overridding works as intended.

Commit:
removed mutagen proxy from arm_content

Commit:
cleaned out all trace

Jbonnett commented:
None of the conflicts need to be merged from my branch, is there a way to just drop them?

tswicegood commented:
Easiest thing to do is merge `armstrong/master` to your `audio_field` branch, then use `git mergetool` to select what you want to do.  You need to commit the changes once you've selected which change wins.

Commit:
Merge github.com:armstrong/armstrong.core.arm_content into audio_field

merged in the changes from the main banch back into the audio field
branch

Conflicts:
	armstrong/core/arm_content/tests/_utils.py
	armstrong/core/arm_content/tests/arm_content_support/admin.py
	armstrong/core/arm_content/tests/arm_content_support/forms.py
	armstrong/core/arm_content/tests/arm_content_support/models.py
	fabfile/__init__.py
	requirements/base.txt
	requirements/dev.txt

Commit:
when i merged I  reintroduced a import i shouldnt have

Commit:
merge required a , where once there was none.

Commit:
pep 8

tswicegood commented on armstrong/core/arm_content/audio/id3reader.py:
```diff
@@ -0,0 +1,657 @@
```
Is there a reason we're copying this over wholesale instead of using the version on PyPI?

Jbonnett replied:
Yes, there are changes. Its MIT licensed per ned himself, so we are permitted to just toss it inside. Its no longer being maintained elsewhere so there is no upstream to contribute changes to. I can toss a copy in my git hub and perhaps package it up a bit, but at the moment I don't think its worth the additional time.

Jbonnett replied:
moved to audio.backends.id3reader

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,122 @@
+        super(AudioFile, self).__init__(*args, **kwargs)
+        if 'metadata' in kwargs:
+            self._metadata = kwargs['metadata']
+        backendKlass = \
```
Please don't use camelCase.  Instead, use `backend_klass` as the style.  Also, `klass` with a `k` is generally only used when you need a variable without any prefix.  This should be `backend_class`.

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,122 @@
+        if 'metadata' in kwargs:
+            self._metadata = kwargs['metadata']
+        backendKlass = \
+            GenericBackend('ARMSTRONG_EXTERNAL_AUDIO_METADATA_BACKEND')\
```
This needs a `defaults` or needs to raise an `ImproperlyConfigured` exception if we can't provide a sane default without some sort of intervention from the developer.

Jbonnett replied:
I figured we would have a facility for specifying defaults to the settings.py I much rather have defaults in the expected location, where they are easily read/changed, than buried in the code.

tswicegood replied:
That's not the model we're using elsewhere in Armstrong.  The idea is to be use Armstrong the way it's meant to be as quickly out of the box as possible with the ability to change things as needed.

tswicegood commented on armstrong/core/arm_content/fields/widgets/audio.py:
```diff
@@ -0,0 +1,14 @@
+
+class AudioFileWidget(ClearableFileInput):
+    def render(self, name, value, attrs):
+        parrent_output = super(AudioFileWidget, self).render(name, value, attrs)
```
Double R here and below on line 14.

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,122 @@
+        raise NotImplementedError
+
+    def render(self, *args, **kwargs):
+        if('armstrong.apps.audio' in settings.INSTALLED_APPS):
```
Why do we need `armstrong.apps.audio` here?

Jbonnett replied:
it calls  the audio app if its installed, to get the js player html for the file.

tswicegood replied:
Why is that in `armstrong.apps.articles` then?  With that there, you can't use the widget without installing our audio app, which isn't the goal.

Jbonnett replied:
This was decided a fair while back and I talked to Arillio rather than you. Perfectly willing to move it inside arm_content, though I do feel as though arm_content has quite a bit jammed in it.

tswicegood replied:
`arm_content` shouldn't have a dependency on something in the `armstrong.apps` namespace.  The dependencies are supposed to go the other way.  Let's move it.

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,122 @@
+        # Model.__init__, see bug #11196.
+        signals.post_init.connect(self.update_metadata, sender=cls)
+
+    def update_metadata(self, instance, force=False, *args, **kwargs):
```
`force` doesn't seem to be getting used...?

Jbonnett replied:
left over from when i was going off the image field code

tswicegood commented on armstrong/core/arm_content/tests/_utils.py:
```diff
@@ -152,3 +158,12 @@ def random_authored_model(klass, *authors):
+
+def load_audio_model(filename, model=AudioModel):
+    #todo: there needs to be a better way to get the filename
+    f = open('./armstrong/core/arm_content/tests/arm_content_support/static/audio/' + filename,"rb+")
```
pep8 requires a space between `,` and the next value I think.

tswicegood commented:
In addition to addressing the comments on the diff, can you also merge in the latest from `armstrong/master`?

Commit:
make load_audio_model generic enough to be called in the audio app tests

Commit:
backend_class(backendKlass) need not exist

removed a unused arg on update metadata

Commit:
i am teh besst spelar

double rr in parent for no good reason

Commit:
Merge github.com:armstrong/armstrong.core.arm_content into audio_field

pulling down changes from master
Conflicts:
	armstrong/core/arm_content/tests/_utils.py

Commit:
killed some needed headers from _utils.py in the merge

tswicegood commented on armstrong/core/arm_content/tests/_utils.py:
```diff
@@ -27,3 +33,12 @@ def random_authored_model(klass, *authors):
+
+
+def load_audio_model(filename, model=AudioModel, model_args={}):
+    f = open(settings.STATIC_ROOT + 'audio/' + filename, "rb+")
```
`STATIC_ROOT` needs to be defined in the settings in the fabfile.

Jbonnett replied:
Yep. got lost in the shuffle somewhere.

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,121 @@
+        if 'metadata' in kwargs:
+            self._metadata = kwargs['metadata']
+        self.backend = \
+            GenericBackend('ARMSTRONG_EXTERNAL_AUDIO_METADATA_BACKEND')\
```
If you don't want to provide a default value here (and if it can't run without at least one of the backends installed, that seems reasonable), we should at least catch the `ImproperlyConfigured` error and provide additional context stating what they need to do (install one of the backends, etc.).

Commit:
updated docs for the audio field

Commit:
moved id3 to its own module.

Commit:
adding the audio widget to the audio field

tswicegood commented on docs/audio.rst:
```diff
@@ -0,0 +1,111 @@
+
+    class Audio(models.Model):
+        audio_file = fields.AudioFileField()
+        #optional fields that audio will pre populate
```
These fields should all have an optional `kwarg` in `AudioFileField.__init__()` with defaults that set these.  See [AuthorsField](https://github.com/armstrong/armstrong.core.arm_content/blob/master/armstrong/core/arm_content/fields/authors.py#L108-109) for an example.

Jbonnett replied:
The audio field will pre-populate the models fields if they are empty as it is now. See the update_metadata function of the audiofield class. What is is that you functionally want changed? Do you want to be able to override the fields representation of them?  As in not audiomodel.artist but audiomodel.audiofield.metadata['artist']?
I wanted to stay away from that because I don't want to deal with saving new versions of the mp3's with different tags, which is the functionality implied by overriding the .metadata of the audio file field, then saving the model. 

if you mean you want to be able to set a.artist='travis' or a=Audiomodel(artist='travis') all of that works as you would expect.

tswicegood replied:
I don't want `artist` to be the only field that this could possibly work with.  The current implementation is inconsistent with the way `AuthorsField` is implemented, where you can explicitly set what the field names are.  The example is:

```
class MyAudioModel(models.Model):
    audio_file = fields.AudioFileField(artist_field_name="custom_artist")
    custom_artist = models.CharField(max_length=100)
```

You can see some example of this in the `AuthorsField` by searching for "customizing" in [the docs](http://armstrong.github.com/armstrong.core.arm_content/authors/).

Jbonnett replied:
currently its any field that shares a name with a piece of metadata. artist is just my example. If you want a mapping i can do that.

Jbonnett replied:
I will write that and get back to you.

Jbonnett replied:
fixed in commit 147d280
also found a bug as a bonus.

Commit:
moved audio backend to apps.audio.backends

Commit:
made the player widget the default for filled ind audio fields

Commit:
no pdb allowed

Commit:
made the kwarg override to audio field attrs generic

Commit:
swapped out the test mp3 for my own

deleted the unused ogg

Commit:
'metadata name'_field_name override kwargs are now accepted

the following is now supported
class MyAudioModel(models.Model):
    audio_file = fields.AudioFileField(artist_field_name = 'performer',
                                       album_field_name = 'album')
    preformer = models.CharField(max_length=100)
    album = models.CharField(max_length=100)

for all the myriad metadata fields

Commit:
pep8

tinio commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,139 @@
+    def _transcode(self, toformat):
+        """
+        returns a transcoded file
+        perhaps should do more complex and diffrent things
```
minor typo: different

tinio commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,139 @@
+from django.forms.fields import FileField as FileFormField
+from django.template.loader import get_template
+from django.template import Context
+from django.conf import settings
```
Is this import still needed?  I don't see settings referenced anywhere else.

tinio commented on armstrong/core/arm_content/fields/widgets/audio.py:
```diff
@@ -0,0 +1,19 @@
+from django.utils.html import escape, conditional_escape
```
Likewise, I don't see escape or conditional_escape used here. Remove imports?

tinio commented on docs/audio.rst:
```diff
@@ -0,0 +1,111 @@
+        });
+        //]]>
+
+This code can be devided into tiwo parts, one part right below //<!CDATA[, which is the conditiontal include of the jplayer.js file, and the actual invocation of the player itself.  The important values to note are the url ``/static/audio/test.mp3``, which is the file url, ``/static/`` which is the static file prefix, ``mp3`` which is the file type.
```
Typos: This code can be **divided** into **two** parts

tinio commented:
@jbonnett Saw a couple of unused imports.  Might want to pass the code through pyflakes or something similar and make sure there aren't others before a merge.

Commit:
pyflakes

Commit:
Code cleanup/docs cleanup

fixed some spelling problems in the docs and added documentation for the
metadata field mapping

Commit:
spelling

tswicegood commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,137 @@
+    def __init__(self, *args, **kwargs):
+        self.overrides = {}
+        for key in kwargs.keys():
+            if '_field_name' in key:
```
This is overly agressive in its matching.  It could match against `_field_names_needed` and pull it out of `kwargs`.  This should be `key.endswith("_field_name")` if we keep this implementation.  Which leads to...

This implementation seems to add to the cyclomatic complexity for no reason.  Do we not know what data is going to be exposed via `metadata`?

Jbonnett replied:
Indeed we do not.  There is a rather extensive list of available tags, just for mp3's. if you use the mutagen backend there are also a ton of tags exposed by the other formats. 

re endswith, i agree, i just couldnt remember what python called it.

Commit:
replaces in with endswith for more precise matching of field name overrides

Commit:
some javascript fixes for the player

Commit:
added a field_name override to the load_audio_file test util

Commit:
added unicode and str functions to the audio field

Commit:
fixes a test that broket from js changes

Commit:
yet more widget js tweaking

Commit:
more messing with the player/widget code

Commit:
forgot to check in the skin for the audio player

dmclain commented on armstrong/core/arm_content/fields/widgets/audio.py:
```diff
@@ -0,0 +1,18 @@
+from django.forms.widgets import ClearableFileInput
+
+
+class AudioFileWidget(ClearableFileInput):
```
If the audio file widget needs the javascript/css at https://github.com/armstrong/armstrong.apps.audio/pull/1/files#L4R15, then it should be declared in the widget (see https://github.com/armstrong/armstrong.hatband/blob/master/armstrong/hatband/widgets/ckeditor.py#L7)

Jbonnett replied:
shiny.

dmclain commented on armstrong/core/arm_content/fields/widgets/audio.py:
```diff
@@ -0,0 +1,18 @@
+
+class AudioFileWidget(ClearableFileInput):
+    def __init__(self, *args, **kwargs):
+        super(AudioFileWidget, self).__init__(*args, **kwargs) 
```
If init is just calling super.init, probably ought to remove the code

Jbonnett replied:
Yeah, at one point it did something else. Now that that has changed, removing it.

dmclain commented on armstrong/core/arm_content/tests/_utils.py:
```diff
@@ -27,3 +33,12 @@ def random_authored_model(klass, *authors):
+
+
+def load_audio_model(filename, file_field_name='audio_file', model=AudioModel, model_args={}):
+    f = open(settings.STATIC_ROOT + 'audio/' + filename, "rb+")
```
Is this file ever closed?

tswicegood replied:
This would be a good spot for `with open(...) as f` so closing and errors and such are automatically handled though I'm not sure what would happen if you return out of a with statement as I've never tracked that.

Jbonnett replied:
its only used in the tests, at the worst its closed when the tests finish. Not that it shouldn't be closed, just that its not the end of the world if it isnt.

dmclain commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,148 @@
+                if key + '_field_name' in self.overrides:
+                    field_name = self.overrides[key + '_field_name']
+                else:
+                    field_name = key
```
`field_name = self.overrides.get(key + '_field_name', key)`

Jbonnett replied:
well thats certainly neater.

Jbonnett replied:
done

dmclain commented on armstrong/core/arm_content/fields/audio.py:
```diff
@@ -0,0 +1,148 @@
+                if hasattr(instance, field_name) and getattr(instance, field_name):
+                        pass
+                elif hasattr(instance, field_name) and audio_file.metadata[key]:
+                    setattr(instance, field_name, audio_file.metadata[key])
```
This logic feels odd. Why isn't it one larger if statement? Why do you only want to call setattr if audio_field.metadata[key] is non false? It seems like as long as you aren't overwriting existing data, there's no harm in allowing 'false' data through. The difference between None, '' and 0 might be significant for somebody in the future, plus it's one less thing to check for

```
if hasattr(instance, field_name) and getattr(instance, field_name) is None:
    setattr(instance, field_name, audio_file.metadata[key])
```

tswicegood replied:
+1 on simplifying, but is that correct?  Would an empty string (assuming that's the default) keep it from ever being set?  Definitely something to write a few test cases around to make sure of the behavior.

Jbonnett replied:
simplified somewhat.

tswicegood commented on armstrong/core/arm_content/models.py:
```diff
@@ -31,7 +31,7 @@ class ContentBase(mixins.AuthorsMixin, mixins.PublicationMixin, models.Model):
     primary_section = models.ForeignKey(Section, null=True)
     sections = models.ManyToManyField(Section, null=True, blank=True,
-            related_name="alternates")
+            related_name="%(class)s_alternates")
```
Not sure I understand the context of this change.  Why is it necessary and what does it have to do with the audio field?

Jbonnett replied:
It has nothing to do with audio, i have a separate pull request open for it. It makes sub classes of Content base not step on each other's sections related name

tswicegood commented on armstrong/core/arm_content/tests/fields/audio.py:
```diff
@@ -0,0 +1,77 @@
+class AudioFieldMetadataTestCase(ArmContentTestCase):
+    def setUp(self):
+        if type(self) is AudioFieldMetadataTestCase:
+            return self.skipTest('parent class')
```
Let's not do testing this way.  It's adding skips which are generally meant to signal that it wasn't able to test something.  When I saw these in the test results I thought it was because I didn't have something configured.

tswicegood replied:
Doh.  Forgot to say how to fix this...  Using `__all__` you can specify what's imported.  Doing that should be enough to ignore this test case.

Jbonnett replied:
done.

tswicegood commented on armstrong/core/arm_content/tests/fields/audio.py:
```diff
@@ -0,0 +1,77 @@
```
Going back through this code, there's an awful lot of functionality in the audio fields code that's only barely touched on (if touched on) by these tests.  Do you think we can get more of the specifics covered in tests?

Commit:
added a Media class to the widget

removed a cruftie __init__ that did nothing

tswicegood commented on armstrong/core/arm_content/templates/audio/player.html:
```diff
@@ -0,0 +1,56 @@
+{% get_static_prefix as STATIC_PREFIX %}
+ 
+<script type="text/javascript">
+    //<![CDATA[
```
We're not XHTML.  We're going to be shipping HTML, so need to CDATA this.  Better yet, put it in an external file.

tswicegood replied:
Correction -- "...so _no_ need to CDATA..."

Jbonnett replied:
done

Commit:
added __all__ to skip the parent class of the audio tests

also fixed a typo and a lack of a import

Commit:
Made the player work with the normal with jQuery or  django.jQuery

Commit:
added a audio mixin

Commit:
audio field will now stuff false values into the model

Commit:
added tests for the audio mixin

made the naming of audio models/fields more consitant across apps.audio and arm_content

Commit:
doc updates to cover the mixin and the diffrent javascript

Commit:
further doc updates for audio

Commit:
removed the class name from the reverse name of sections on BaseContent

also removed a unused template tag load from the player
